### ISSUE TITLE
Support query cancellation

### ISSUE THREAD

**Issue by whitelynx** [tgriesser/knex#707]
Not sure about the other database drivers, but node-postgres has [support for cancelling running queries](https://github.com/brianc/node-postgres/wiki/pg#cancelconfig-client-query). It would be nice to utilize [bluebird's cancellation logic](https://github.com/petkaantonov/bluebird/blob/master/API.md#cancellation) to allow queries to be cancelled via the promise mechanism.

**Comment by whitelynx** [comment]
If that's not feasible at the moment, is there a way to access the node-postgres `client` and `query` objects so the query could be cancelled by user code?

**Comment by tgriesser** [comment]
Curious, what sorts of situations are you looking to cancel queries - just so I understand the use case a little better... would it be based on some sort of timeout (query taking too long) or something else?

**Comment by whitelynx** [comment]
Generally speaking, it'd be in response to a timeout; however, I've been looking at switching [my web-based PostgreSQL client](https://github.com/whitelynx/web-pgq) over to using knex, and in that case it'd be in response to a button press on the client side.

**Comment by whitelynx** [comment]
Just curious... If I were to try and implement this, do you have any guidelines for adding a DB client API function that may only be implemented by some clients? I was thinking of at least implementing this in the PostgreSQL client, and we can work on adding it to other clients later.

**Comment by elhigu** [comment]
There are no guidelines how to do this. 

Would be good to study which dialects could support the feature and then consider if it should be implemented in knex or separate library. 

API should be designed to be generally usable, but if some dialect does not support it using the feature should throw an exception...

**Comment by whitelynx** [comment]
In [the Bluebird docs on Cancellation](http://bluebirdjs.com/docs/api/cancellation.html), where they describe the differences between the old 2.x cancellation semantics and the new 3.x approach, they say:

> The new cancellation has "don't care" semantics while the old cancellation had abort semantics. Cancelling a promise simply means that its handler callbacks will not be called.

_..._

> Note that the `onCancel` hook is really an optional disconnected optimization, there is no real requirement to register any cancellation hooks for cancellation to work. As such, any errors that may occur while inside the `onCancel` callback are not caught and turned into rejections.

_..._

> As shown in the example the handlers registered with `.finally` are called even if the promise is cancelled. Another such exception is [.reflect()](http://bluebirdjs.com/docs/api/reflect.html). No other types of handlers will be called in case of cancellation. This means that in `.then(onSuccess, onFailure)` neither `onSuccess` or `onFailure` handler is called. This is similar to how [`Generator#return`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return) works - only active `finally` blocks are executed and then the generator exits.

Since errors from `onCancel` are not integrated into the promise chain, it makes sense that no error should be thrown by backends that do not implement cancellation. (they simply do not register an `onCancel` handler - the user can still take advantage of the other cancellation semantics built into Bluebird; they just won't get the query itself canceled, since that's an "optional disconnected optimization")

I also suggest that cancellation would still need to be enabled _by the calling code_; I do _not_ think knex should do this. On knex's end, this is pretty simple to achieve; any backend that implements cancellation would just check if the `onCancel` function was passed into the `Promise` constructor, and only register its handler if that was the case.

As far as whether this should be in knex itself or in a separate library... I'll have to do more research on the driver libraries for the various dialects in order to know whether or not they support query cancellation. If it is decided that this should be in a separate library, how should that be achieved? Are there any sort of extension points built into knex's dialects I could leverage to allow me to replace the promise returned by running the query?

**Comment by mightyguava** [comment]
@wubzz @byronmwong following up on this issue. It looks like implementing query cancellation for PostgreSQL and other drivers should be pretty easy now that @byronmwong added it for MySQL. Is there any blockers here or is it just waiting for a PR from someone?

**Comment by mohannad-musleh** [comment]
Is it possible to cancel the query manually?

for context:
I'm building a desktop application using electron js, and i need to give the user the capability to cancel the running query

**Comment by jtlapp** [comment]
Gosh, I'm surprised there isn't a way to cancel a query that's timing out. @mohannad-musleh, I'm also facing this building an Electron app. For now, my workaround will be to shorten the timeout and repeatedly give the user the option to try again.

**Comment by kibertoad** [comment]
It should be possible to add one. We already invoke query cancellation on timeout, guess it should be possible to expose manual cancellation. Is there anyone willing to work on a PR (with some guidance)?

**Comment by irfancnk** [comment]
Still an issue.
I am also looking for cancelling a pending migration. For example when developer accidently runs a migration that does not return a Promise, runs into a situation where migration stays on Pending, probably forever?

**Comment by kibertoad** [comment]
@solanoepalacio I am one of the maintainers, yes. Would be happy to help!

**Comment by jhiesey** [comment]
@solanoepalacio @kibertoad I'm also interested in implementing this, potentially as soon as this week.

What I have in mind is a chainable method on the query object, tentatively named `abortOnSignal`. This method takes an [AbortSignal](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal) (available since Node v15). AbortController/AbortSignal came out of an extensive discussion about the right interface for cancelling HTTP requests made using the `fetch` API.

As an example, you could implement your own query timeouts (without using knex's built-in `query.timeout method`) like this:
```js
async function runMyQueryWithTimeout() {
  const abortController = new AbortController()
  setTimeout(() => abortController.abort(), 1000)

  return await knex('my_table')
    .select('id', 'my_column')
    .abortOnSignal(abortController.signal)
}
```

The promise returned by `runMyQueryWithTimeout()` would either resolve to the query result, or reject with an `AbortError`.

**Comment by JakeHamix** [comment]
I am currently facing this issue as well. Some of the queries take a long time to complete, so users are tempted to close/duplicate application windows or move through the application unhindered. The HTTP requests are aborted, however, the database queries do not cancel themselves - which tends to create massive CPU consumption for no reason. The solution with AbortSignal would be a very welcome sight, I will create a placeholder solution until then.

I have partially solved this in the past. We needed to see long-running queries in the application GUI and the ability to cancel them - the solution was a method, which cancels the request by **__knexUid.** There is added complexity if your application scales backends - nevertheless, this is just a MSSQL solution:

```javascript
                const activeRequests = ourGlobalProcessObject.knex.client.pool.used.map(({ resource: { __knexUid, request } }) => {
                    return {
                        __knexUid,
                        request,
                    };
                });

                const requestToDestroy = _.find(activeRequests, { __knexUid: requestId });
                if (requestToDestroy) {
                    requestToDestroy.request.cancel();
                    return true;
                }
```

**Comment by exoRift** [comment]
Any status updates on this?

**Comment by mercmobily** [comment]
@jhiesey @solanoepalacio Are you still interested in implementing this?

**Comment by mercmobily** [comment]
This issue (closed as duplicate) has also some information and a code example: https://github.com/knex/knex/issues/4939

**Comment by mercmobily** [comment]
(I managed to work on this on a tab that was open BEFORE the PR was submitted... hence the rewrite of the response)

This feature has now been implemented: PR #6253, submitted by @exoRift with assistance from GitHub Copilot, implements exactly the AbortSignal-based approach that @jhiesey proposed. The PR adds an `abortOnSignal()` method to both QueryBuilder and Raw queries, allowing manual query cancellation using the standard AbortController/AbortSignal pattern.

The implementation in PR #6253 properly leverages the existing cancellation infrastructure that's already used for timeout-based cancellation. It includes comprehensive tests, TypeScript definitions, and documentation. The PR correctly handles edge cases such as already-aborted signals, invalid signal parameters, and compatibility with existing timeout functionality. It requires Node.js 15 or higher for AbortSignal support and currently works with MySQL and PostgreSQL drivers.

This PR effectively resolves this long-standing issue by providing a clean, modern API for manual query cancellation that aligns with JavaScript standards. The implementation follows the exact pattern discussed in this issue thread, making it possible to cancel queries in response to user actions, navigation events, or any other programmatic triggers.

However, it would be valuable to verify that the PR works correctly with all drivers that support cancellation. Currently, only three drivers have `canCancelQuery` set to true:
- PostgreSQL (lib/dialects/postgres/index.js:301)
- PostgreSQL Native (lib/dialects/pgnative/index.js:9)
- MySQL/MySQL2 (lib/dialects/mysql/index.js:203)

CockroachDB extends the PostgreSQL client and implements its own `cancelQuery()` method (lib/dialects/cockroachdb/index.js:43), but it should inherit the `canCancelQuery: true` property from PostgreSQL. The PR should be tested to ensure it works correctly with all these drivers.

Other drivers (SQLite3, Better-SQLite3, MSSQL, Oracle, OracleDB, and Redshift) do not currently support query cancellation in Knex. For MSSQL specifically, whilst the driver doesn't have built-in support, the community has demonstrated working implementations (as shown by @JakeHamix in #4939) using the tedious driver's request cancellation capabilities. This could potentially be adapted to provide official MSSQL support in a future enhancement.

**Related issues and PRs:**
- #6253 - **PR implementing AbortSignal-based query cancellation (resolves this issue)**
- #4939 - Duplicate issue with additional MSSQL implementation details
- #3666 - Fixed unresolved promise in cancelQuery
- #3518 - PostgreSQL timeout using cancel instead of terminate
- #2636 - Kill queries after timeout for PostgreSQL
- #1454 -Original MySQL killQuery implementation

**Comment by jhiesey** [comment]
> [@jhiesey](https://github.com/jhiesey) [@solanoepalacio](https://github.com/solanoepalacio) Are you still interested in implementing this?

Nope, I got distracted with other things and haven't had time to get back to it
